include/cuda/std/memory_resource

```diff
@@ -0,0 +1,268 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include "cstddef"
+#include "detail/__config"
+#include "detail/__pragma_push"
+#include "stream"
+#include "version"
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+/**
+ * @brief Specifies the kind of memory of an allocation.
+ *
+ * Memory allocation kind determines where memory can be accessed and the
+ * performance characteristics of accesses.
+ *
+ */
+enum class memory_kind {
+  device,  ///< Device memory accessible only from device
+  unified, ///< Unified memory accessible from both host and device
+  pinned,  ///< Page-locked system memory accessible from both host and device
+  host     ///< System memory only accessible from host code
+};
+
+/**
+ * @brief Abstract interface for synchronous memory allocation.
+ *
+ * @tparam _Kind The `memory_kind` of the allocated memory.
+ */
+template <memory_kind _Kind> class memory_resource {
```

Comment:
Conditionally inherit from `std::pmr::memory_resource` based on language version _and_ `memory_kind`?

Reply:
What about `cuda::std::pmr::memory_resource`? That would require porting libc++'s `<memory_resource>` header.